File: apps/ios/App/Charts/FrictionCircle.swift
import SwiftUI
import TrackEvolutionKit

/// The friction circle (#186) — the counterpart of `frictionCircleSvg` and
/// `gripReadoutHtml` in `public/js/grip.js`.
///
/// Every 20 m sample of the highlighted laps plotted lateral against
/// longitudinal G, over a dim envelope of the session's other laps, with a
/// dashed reference arc at the session's own peak combined G. A driver who
/// brakes in a straight line, turns, then accelerates draws a cross; one who
/// trails the brake in and feeds the power out fills the circle, and the empty
/// space between the two is the lost time. The maths is the Kit's `Grip`,
/// pinned to the web implementation by `contracts/logic/grip.json`; this file
/// draws.
///
/// Three things about the drawing are load-bearing.
///
/// **Braking is up.** `longG` is negative under braking, so the y mapping adds
/// rather than subtracts — deceleration is what a driver feels pitching the car
/// forward, and it is the one axis on this chart with a fixed place in their
/// head. Same call the web made, and `GripTests` is not where it would be
/// caught, so it is stated here.
///
/// **The plot area is square** (`aspectRatio(1, contentMode: .fit)`), because a
/// circle that draws as an ellipse makes the whole reading wrong.
///
/// **Hand-rolled on `Canvas`, not Swift Charts.** This is a polar picture —
/// concentric rings, a dashed arc, thousands of unconnected points — and a
/// `PointMark` per sample inside the panel's sheet is exactly the "chart of this
/// many marks never settles" hazard `LapChannelChart` documents. The points are
/// drawn as two paths (dim envelope, then the lit laps), which is also what
/// keeps a full session cheap: `MAX_TOTAL_VALUES` caps a stored session's
/// channels, so `latG` tops out in the low thousands of samples.
///
/// **A tap points, and a pointer borrows.** The web answers a *hover* over a dot
/// by marking that distance on every chart and ringing the place on the track
/// map; a phone has no pointer, so NS-34 ticket 3 gave that to a tap instead.
/// An iPad with a trackpad has both, so ticket 5 wires the pointer up as well —
/// additive to the tap, never a replacement, because the same iPad is used by
/// touch a moment later. A hovered mark lasts as long as the pointer is over the
/// plot; a tapped one stays until it is tapped away.
struct FrictionCircle: View {
    let channels: SessionChannels
    /// Channel-lap indexes in slot order, as `LapChannelPanel` keeps them.
    let lit: [Int]
    let slots: [Color]
    /// The session's own lap number for a channel entry.
    let lapNumber: (Int) -> Int
    /// Tapping a sample answers "which corner is this dot" on everything drawn
    /// beside it (NS-34 ticket 3). Nil clears the answer.
    var onHit: (ChannelHit?) -> Void = { _ in }
    /// The same answer from a *pointer* (NS-34 ticket 5), which the panel holds
    /// only while the pointer is over the plot. Nil means it has left.
    var onHover: (ChannelHit?) -> Void = { _ in }
    /// Whether to draw the laps that aren't highlighted (⌘F).
    ///
    /// The envelope is what a lit lap is read *against*, so it is on by default
    /// and only a keyboard can turn it off — on a crowded session the dim cloud
    /// can bury the three laps being compared, and that is the one case worth a
    /// shortcut rather than a control.
    var showEnvelope: Bool = true

    /// Ring spacing, in G. 0.5 G rings are readable on every car this app sees;
    /// a slow car simply draws fewer of them.
    private static let ringStepG: Double = 0.5

    private struct Row: Identifiable {
        let color: Color
        let lap: Grip.LapShares
        var id: Int { lap.chIdx }
    }

    var body: some View {
        if let sg = Grip.sessionGrip(channels) {
            let rows = lit.enumerated().compactMap { slot, chIdx -> Row? in
                guard slot < slots.count, let lap = sg.laps.first(where: { $0.chIdx == chIdx }) else { return nil }
                return Row(color: slots[slot], lap: lap)
            }
            TECard {
                VStack(alignment: .leading, spacing: 10) {
                    VStack(alignment: .leading, spacing: 2) {
                        Text("Friction circle")
                            .teStyle(.eyebrow)
                            .foregroundStyle(Color(.textMuted))
                        Text("How much of the tyre is being used. 20 m samples, so this is the shape of grip usage, not peak G.")
                            .teStyle(.xxs)
                            .foregroundStyle(Color(.textFaint))
                    }
                    Canvas { context, size in
                        draw(sg, in: context, size: size)
                    }
                    .aspectRatio(1, contentMode: .fit)
                    .frame(maxWidth: .infinity)
                    // Tap *and* hover, over one hit-test: a finger commits a
                    // mark and a pointer borrows one, and both ask the same
                    // question of the same geometry so the two can never come to
                    // different answers about which dot is under them.
                    .overlay {
                        GeometryReader { geometry in
                            Color.clear
                                .contentShape(Rectangle())
                                .onTapGesture { location in
                                    onHit(hit(sg, at: location, size: geometry.size))
                                }
                                .onContinuousHover { phase in
                                    switch phase {
                                    case .active(let location):
                                        onHover(hit(sg, at: location, size: geometry.size))
                                    case .ended:
                                        onHover(nil)
                                    @unknown default:
                                        onHover(nil)
                                    }
                                }
                        }
                    }
                    readout(sg, rows)
                }
            }
            .accessibilityElement(children: .ignore)
            .accessibilityLabel("Friction circle: cornering against braking and acceleration")
            .accessibilityValue(summary(sg, rows))
            .accessibilityIdentifier("frictionCircle")
        }
    }

    // MARK: - Plot

    /// The axis domain: the furthest sample, or the reference arc if it somehow
    /// reaches further, with a little air — rounded up to a ring boundary so the
    /// outermost ring is a labelled one. `axisMaxG` in the JS.
    private func axisMaxG(_ sg: Grip.SessionGrip) -> Double {
        let need = max(sg.maxG, sg.peakG ?? 0) * 1.04
        return max(Self.ringStepG, (need / Self.ringStepG).rounded(.up) * Self.ringStepG)
    }

    /// The plotted sample nearest a tap, among the **highlighted** laps only.
    ///
    /// The envelope behind them is every other lap at once; a tap landing on it
    /// has no single answer to give, so it clears instead. Within a lit lap the
    /// nearest sample wins, but only inside a finger-sized radius — a tap on
    /// empty canvas should mean "never mind", not "the closest dot, wherever it
    /// is".
    ///
    /// The transform is `draw`'s, deliberately duplicated rather than hoisted:
    /// the drawing takes a `GraphicsContext` and this takes a tap, and pulling
    /// the four lines they share into a shape object would be more indirection
    /// than the arithmetic is worth. They fail together — a change to one that
    /// misses the other puts the ring in the wrong corner, which is visible the
    /// first time it is used.
    private func hit(_ sg: Grip.SessionGrip, at location: CGPoint, size: CGSize) -> ChannelHit? {
        let inset: CGFloat = 18
        let side = min(size.width, size.height) - inset * 2
        guard side > 0 else { return nil }
        let center = CGPoint(x: size.width / 2, y: size.height / 2)
        let scale = Double(side / 2) / axisMaxG(sg)
        guard scale > 0 else { return nil }

        var best: (distance: CGFloat, hit: ChannelHit)?
        for chIdx in lit {
            guard let lap = Grip.gripLaps(channels).first(where: { $0.chIdx == chIdx }) else { continue }
            let points = Grip.gripPoints(lap.entry)
            guard points.count > 1 else { continue }
            for p in points {
                let plotted = CGPoint(
                    x: center.x + CGFloat(p.lat * scale),
                    y: center.y + CGFloat(p.long * scale)
                )
                let d = hypot(plotted.x - location.x, plotted.y - location.y)
                if best == nil || d < best!.distance {
                    best = (
                        d,
                        ChannelHit(
                            chIdx: chIdx,
                            k: p.k,
                            frac: points.count > 1 ? Double(p.k) / Double(points.count - 1) : 0
                        )
                    )
                }
            }
        }
        guard let best, best.distance <= 22 else { return nil }
        return best.hit
    }

    private func draw(_ sg: Grip.SessionGrip, in context: GraphicsContext, size: CGSize) {
        // Room for the axis words above and below the plot.
        let inset: CGFloat = 18
        let side = min(size.width, size.height) - inset * 2
        guard side > 0 else { return }
        let center = CGPoint(x: size.width / 2, y: size.height / 2)
        let axis = axisMaxG(sg)
        let scale = Double(side / 2) / axis
        // Braking is up: longG is negative under braking, so y *adds*.
        func point(_ lat: Double, _ long: Double) -> CGPoint {
            CGPoint(x: center.x + CGFloat(lat * scale), y: center.y + CGFloat(long * scale))
        }

        // Rings every ringStepG, labelled along the +x axis.
        var ring = Self.ringStepG
        while ring <= axis + 1e-9 {
            let r = CGFloat(ring * scale)
            context.stroke(
                Path(ellipseIn: CGRect(x: center.x - r, y: center.y - r, width: r * 2, height: r * 2)),
                with: .color(Color(.chartGrid)), lineWidth: 1
            )
            context.draw(
                context.resolve(
                    Text(String(format: "%.1f", ring))
                        .font(.system(size: 9).monospacedDigit())
                        .foregroundColor(Color(.textFaint))
                ),
                at: CGPoint(x: center.x + r, y: center.y + 9)
            )
            ring += Self.ringStepG
        }

        // The axes.
        var axes = Path()
        axes.move(to: CGPoint(x: center.x - CGFloat(side / 2), y: center.y))
        axes.addLine(to: CGPoint(x: center.x + CGFloat(side / 2), y: center.y))
        axes.move(to: CGPoint(x: center.x, y: center.y - CGFloat(side / 2)))
        axes.addLine(to: CGPoint(x: center.x, y: center.y + CGFloat(side / 2)))
        context.stroke(axes, with: .color(Color(.borderStrong)), lineWidth: 1)

        // The dim envelope: every lap that isn't highlighted, as one path.
        let litSet = Set(lit)
        var dim = Path()
        var hot: [(Color, Path)] = []
        for lap in Grip.gripLaps(channels) {
            let points = Grip.gripPoints(lap.entry)
            if let slot = lit.firstIndex(of: lap.chIdx), slot < slots.count {
                var path = Path()
                for p in points {
                    let c = point(p.lat, p.long)
                    path.addEllipse(in: CGRect(x: c.x - 2.4, y: c.y - 2.4, width: 4.8, height: 4.8))
                }
                hot.append((slots[slot], path))
            } else if showEnvelope, !litSet.contains(lap.chIdx) {
                for p in points {
                    let c = point(p.lat, p.long)
                    dim.addEllipse(in: CGRect(x: c.x - 1.3, y: c.y - 1.3, width: 2.6, height: 2.6))
                }
            }
        }
        context.fill(dim, with: .color(Color(.chartDim)))

        // The reference arc: what this car did today, at the 99th percentile.
        if let peak = sg.peakG, peak > 0 {
            let r = CGFloat(peak * scale)
            context.stroke(
                Path(ellipseIn: CGRect(x: center.x - r, y: center.y - r, width: r * 2, height: r * 2)),
                with: .color(Color(.accent)),
                style: StrokeStyle(lineWidth: 1.5, dash: [5, 4])
            )
            context.draw(
                context.resolve(
                    Text(String(format: "%.2f G", peak))
                        .font(.system(size: 10, weight: .semibold).monospacedDigit())
                        .foregroundColor(Color(.accentInk))
                ),
                at: CGPoint(x: center.x, y: center.y - r - 8)
            )
        }

        for (color, path) in hot {
            context.fill(path, with: .color(color.opacity(0.75)))
        }

        // Axis words. The sides say "cornering" rather than left/right, because
        // the side is derived from the steering sign rather than stored.
        func label(_ text: String, at position: CGPoint, size fontSize: CGFloat = 10) {
            context.draw(
                context.resolve(Text(text).font(.system(size: fontSize)).foregroundColor(Color(.textFaint))),
                at: position
            )
        }
        label("braking", at: CGPoint(x: center.x, y: center.y - CGFloat(side / 2) - 9))
        label("power", at: CGPoint(x: center.x, y: center.y + CGFloat(side / 2) + 9))
        // The plot is a disc, so a corner of the square is the one place no
        // sample can ever land.
        context.draw(
            context.resolve(Text("cornering (G)").font(.system(size: 10)).foregroundColor(Color(.textFaint))),
            in: CGRect(
                x: center.x - CGFloat(side / 2), y: center.y - CGFloat(side / 2),
                width: 100, height: 14
            )
        )
    }

    // MARK: - Read-out

    /// The share of *loaded* samples spent doing two things at once, per
    /// highlighted lap, with the session pooled underneath once there is more
    /// than one lap to pool — `gripReadoutHtml` in the JS.
    @ViewBuilder
    private func readout(_ sg: Grip.SessionGrip, _ rows: [Row]) -> some View {
        if !rows.isEmpty {
            VStack(alignment: .leading, spacing: 8) {
                Grid(alignment: .trailing, horizontalSpacing: 10, verticalSpacing: 8) {
                    GridRow {
                        Text("")
                            .gridColumnAlignment(.leading)
                        Text("Braking +\ncornering")
                            .teStyle(.eyebrow)
                            .multilineTextAlignment(.trailing)
                            .foregroundStyle(Color(.textFaint))
                        Text("Cornering +\npower")
                            .teStyle(.eyebrow)
                            .multilineTextAlignment(.trailing)
                            .foregroundStyle(Color(.textFaint))
                    }
                    ForEach(rows) { row in
                        GridRow {
                            HStack(spacing: 6) {
                                Circle().fill(row.color).frame(width: 8, height: 8)
                                Text("Lap \(String(lapNumber(row.lap.chIdx)))")
                                    .teStyle(.xs)
                                    .foregroundStyle(Color(.textMuted))
                            }
                            .gridColumnAlignment(.leading)
                            percent(row.lap.trailPct, color: Color(.textStrong))
                            percent(row.lap.powerPct, color: Color(.textStrong))
                        }
                    }
                    if sg.laps.count >= 2 {
                        GridRow {
                            Text("Session")
                                .teStyle(.xs)
                                .italic()
                                .foregroundStyle(Color(.textMuted))
                                .gridColumnAlignment(.leading)
                            percent(sg.all.trailPct, color: Color(.textMuted))
                            percent(sg.all.powerPct, color: Color(.textMuted))
                        }
                    }
                }
                Text("Share of the samples where the tyre was working (\(String(format: "%.2f", Grip.MIN_LOAD_G)) G combined or more) spent doing two things at once. Brake straight, turn, then accelerate and both stay low — the gap to the arc is time.")
                    .teStyle(.xxs)
                    .foregroundStyle(Color(.textFaint))
            }
        }
    }

    private func percent(_ value: Double, color: Color) -> some View {
        Text("\(Int(value.rounded()))%")
            .teStyle(.sm)
            .monospacedDigit()
            .foregroundStyle(color)
    }

    /// What VoiceOver gets: the two figures per lap and the envelope — a scatter
    /// is exactly the kind of picture a screen-reader user cannot see.
    private func summary(_ sg: Grip.SessionGrip, _ rows: [Row]) -> String {
        var parts = rows.map { row in
            "Lap \(String(lapNumber(row.lap.chIdx))): \(Int(row.lap.trailPct.rounded())) percent braking while cornering, "
                + "\(Int(row.lap.powerPct.rounded())) percent cornering on the power"
        }
        if let peak = sg.peakG {
            parts.append("Peak combined \(String(format: "%.2f", peak)) G")
        }
        // The envelope going away changes what the plot shows, and a scatter is
        // the one picture a screen-reader user can't check for themselves — so
        // ⌘F says so here rather than only on screen.
        if !showEnvelope {
            parts.append("Other laps hidden")
        }
        return parts.joined(separator: ". ")
    }
}

File: apps/android/app/src/main/kotlin/app/trackevolution/ui/charts/FrictionCircle.kt
package app.trackevolution.ui.charts

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.PointMode
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.testTag
import androidx.compose.ui.text.TextMeasurer
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.drawText
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.rememberTextMeasurer
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import app.trackevolution.core.Grip
import app.trackevolution.core.model.SessionChannels
import app.trackevolution.ui.theme.TrackCard
import app.trackevolution.ui.theme.TrackTheme
import kotlin.math.ceil
import kotlin.math.hypot
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/** Ring spacing, in G. A slow car simply draws fewer rings. */
private const val RING_STEP_G = 0.5

/**
 * The friction circle (#186) — the counterpart of `frictionCircleSvg` and
 * `gripReadoutHtml` in `public/js/grip.js`.
 *
 * Every 20 m sample of the highlighted laps plotted lateral against
 * longitudinal G, over a dim envelope of the session's other laps, with a
 * dashed reference arc at the session's own peak combined G. A driver who
 * brakes in a straight line, turns, then accelerates draws a cross; one who
 * trails the brake in and feeds the power out fills the circle, and the empty
 * space between the two is the lost time. The maths is `Grip` in `:core`,
 * pinned to the web implementation by `contracts/logic/grip.json`; this file
 * draws.
 *
 * Three things about the drawing are load-bearing.
 *
 * **Braking is up.** `longG` is negative under braking, so the y mapping adds
 * rather than subtracts — deceleration is what a driver feels pitching the car
 * forward, and it is the one axis here with a fixed place in their head. A
 * Compose canvas already grows y downward, so the JS mapping ports across
 * unchanged; inverting it "to fix the sign" draws the chart upside down and
 * fails no test, which is why it is written down in both ports.
 *
 * **The plot area is square** (`aspectRatio(1f)`), because a circle that draws
 * as an ellipse makes the whole reading wrong.
 *
 * **The samples go through `drawPoints`, not a `drawCircle` each.** A stored
 * session's channels are capped by `MAX_TOTAL_VALUES`, so `latG` runs to a few
 * thousand samples — one batched call per group, rather than one draw call per
 * point.
 *
 * There is no *hover* — a phone has no pointer — but there is a **tap**, and it
 * does what the web's hover does: [onHit] answers "which corner is this dot" on
 * everything drawn beside the panel (NS-34 ticket 3). That was worth nothing
 * while the panel lived inside a session card with the track map a long scroll
 * away; it is worth having now the two sit side by side at expanded width. The
 * read-out under the plot still carries the meaning when nothing is tapped.
 */
@Composable
fun FrictionCircle(
    channels: SessionChannels,
    /** Channel-lap indexes in slot order, as [LapChannelChart] keeps them. */
    lit: List<Int>,
    slots: List<Color>,
    /** The session's own lap number for a channel entry. */
    lapNumber: (Int) -> Int,
    modifier: Modifier = Modifier,
    /**
     * Tapping a sample answers "which corner is this dot" on everything drawn
     * beside it (NS-34 ticket 3). Null clears the answer.
     */
    onHit: (ChannelHit?) -> Unit = {},
) {
    val sg = remember(channels) { Grip.sessionGrip(channels) } ?: return
    val gripLaps = remember(channels) { Grip.gripLaps(channels) }
    val colors = TrackTheme.colors
    val measurer = rememberTextMeasurer()

    val rows = lit.mapIndexedNotNull { slot, chIdx ->
        sg.laps.firstOrNull { it.chIdx == chIdx }?.let { slots[slot % slots.size] to it }
    }

    val summary = buildString {
        rows.forEachIndexed { i, (_, lap) ->
            if (i > 0) append(". ")
            append("Lap ${lapNumber(lap.chIdx)}: ${lap.trailPct.roundToInt()} percent braking while ")
            append("cornering, ${lap.powerPct.roundToInt()} percent cornering on the power")
        }
        sg.peakG?.let { append(". Peak combined ${"%.2f".format(it)} G") }
    }

    TrackCard(
        modifier
            .fillMaxWidth()
            .clearAndSetSemantics {
                testTag = "frictionCircle"
                contentDescription = "Friction circle: cornering against braking and acceleration. $summary"
            },
    ) {
        Column(Modifier.padding(12.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
            Text("Friction circle", style = TrackTheme.typography.xxs, color = colors.textMuted)
            Text(
                "How much of the tyre is being used. 20 m samples, so this is the shape of " +
                    "grip usage, not peak G.",
                style = TrackTheme.typography.xxs,
                color = colors.textFaint,
            )
            val faint = colors.textFaint
            val grid = colors.chartGrid
            val border = colors.borderStrong
            val accent = colors.accent
            val dim = colors.chartDim
            val labelStyle = TrackTheme.typography.xxs.copy(color = faint)
            val arcStyle = TrackTheme.typography.xxs.copy(color = colors.accentInk)
            Canvas(
                Modifier
                    .fillMaxWidth()
                    .aspectRatio(1f)
                    // Tap, not hover: a phone has no pointer, and taking the
                    // web's hover behaviour back is the point of NS-34 ticket 3.
                    // A mouse on a tablet or Chromebook is ticket 5's, and is
                    // additive to this rather than a replacement.
                    .pointerInput(sg, lit, gripLaps) {
                        detectTapGestures { offset ->
                            onHit(hitAt(offset, size.width, size.height, sg, gripLaps, lit, density))
                        }
                    },
            ) {
                drawPlot(
                    sg = sg,
                    gripLaps = gripLaps,
                    lit = lit,
                    slots = slots,
                    grid = grid,
                    border = border,
                    accent = accent,
                    dim = dim,
                    measurer = measurer,
                    labelStyle = labelStyle,
                    arcStyle = arcStyle,
                )
            }
            Readout(sg = sg, rows = rows, lapNumber = lapNumber)
        }
    }
}

/**
 * The axis domain: the furthest sample, or the reference arc if it somehow
 * reaches further, with a little air — rounded up to a ring boundary so the
 * outermost ring is a labelled one. `axisMaxG` in the JS.
 */
private fun axisMaxG(sg: Grip.SessionGrip): Double {
    val need = max(sg.maxG, sg.peakG ?: 0.0) * 1.04
    return max(RING_STEP_G, ceil(need / RING_STEP_G) * RING_STEP_G)
}

/**
 * The plotted sample nearest a tap, among the **highlighted** laps only.
 *
 * The envelope behind them is every other lap at once; a tap landing on it has no
 * single answer to give, so it clears instead. Within a lit lap the nearest
 * sample wins, but only inside a finger-sized radius — a tap on empty canvas
 * should mean "never mind", not "the closest dot, wherever it is".
 *
 * The transform is [drawPlot]'s, deliberately duplicated rather than hoisted: one
 * takes a `DrawScope` and this takes a tap, and pulling the four lines they share
 * into a shape object would be more indirection than the arithmetic is worth.
 * They fail together, and visibly — a change to one that misses the other puts
 * the ring in the wrong corner the first time it is used.
 */
internal fun hitAt(
    offset: Offset,
    width: Int,
    height: Int,
    sg: Grip.SessionGrip,
    gripLaps: List<Grip.GripLap>,
    lit: List<Int>,
    density: Float,
): ChannelHit? {
    val inset = 18f * density
    val side = min(width.toFloat(), height.toFloat()) - inset * 2
    if (side <= 0f) return null
    val cx = width / 2f
    val cy = height / 2f
    val scale = (side / 2) / axisMaxG(sg).toFloat()
    if (scale <= 0f) return null

    var bestDistance = Float.MAX_VALUE
    var best: ChannelHit? = null
    for (chIdx in lit) {
        val lap = gripLaps.firstOrNull { it.chIdx == chIdx } ?: continue
        val points = Grip.gripPoints(lap.entry)
        if (points.size < 2) continue
        for (p in points) {
            val x = cx + (p.lat * scale).toFloat()
            val y = cy + (p.long * scale).toFloat()
            val d = hypot(x - offset.x, y - offset.y)
            if (d < bestDistance) {
                bestDistance = d
                best = ChannelHit(chIdx = chIdx, k = p.k, frac = p.k.toDouble() / (points.size - 1))
            }
        }
    }
    return if (bestDistance <= 22f * density) best else null
}

private fun DrawScope.drawPlot(
    sg: Grip.SessionGrip,
    gripLaps: List<Grip.GripLap>,
    lit: List<Int>,
    slots: List<Color>,
    grid: Color,
    border: Color,
    accent: Color,
    dim: Color,
    measurer: TextMeasurer,
    labelStyle: TextStyle,
    arcStyle: TextStyle,
) {
    // Room for the axis words above and below the plot.
    val inset = 18f * density
    val side = min(size.width, size.height) - inset * 2
    if (side <= 0f) return
    val cx = size.width / 2
    val cy = size.height / 2
    val axis = axisMaxG(sg)
    val scale = (side / 2) / axis.toFloat()

    // Braking is up: longG is negative under braking, and a canvas grows y
    // downward, so the value is *added* — see the file's documentation.
    fun point(lat: Double, long: Double) = Offset(cx + (lat * scale).toFloat(), cy + (long * scale).toFloat())

    // Rings every RING_STEP_G, labelled along the +x axis.
    var ring = RING_STEP_G
    while (ring <= axis + 1e-9) {
        val r = (ring * scale).toFloat()
        drawCircle(color = grid, radius = r, center = Offset(cx, cy), style = Stroke(width = 1f * density))
        val label = measurer.measure("%.1f".format(ring), labelStyle)
        drawText(label, topLeft = Offset(cx + r - label.size.width / 2f, cy + 3f * density))
        ring += RING_STEP_G
    }

    // The axes.
    drawLine(border, Offset(cx - side / 2, cy), Offset(cx + side / 2, cy), strokeWidth = 1f * density)
    drawLine(border, Offset(cx, cy - side / 2), Offset(cx, cy + side / 2), strokeWidth = 1f * density)

    // The dim envelope: every lap that isn't highlighted, batched into one call.
    val dimPoints = ArrayList<Offset>()
    val hot = ArrayList<Pair<Color, List<Offset>>>()
    for (lap in gripLaps) {
        val points = Grip.gripPoints(lap.entry).map { point(it.lat, it.long) }
        val slot = lit.indexOf(lap.chIdx)
        if (slot >= 0) hot.add(slots[slot % slots.size] to points) else dimPoints.addAll(points)
    }
    if (dimPoints.isNotEmpty()) {
        drawPoints(
            points = dimPoints,
            pointMode = PointMode.Points,
            color = dim,
            strokeWidth = 2.6f * density,
            cap = StrokeCap.Round,
        )
    }

    // The reference arc: what this car did today, at the 99th percentile.
    val peak = sg.peakG
    if (peak != null && peak > 0) {
        val r = (peak * scale).toFloat()
        drawCircle(
            color = accent,
            radius = r,
            center = Offset(cx, cy),
            style = Stroke(
                width = 1.5f * density,
                pathEffect = PathEffect.dashPathEffect(floatArrayOf(5f * density, 4f * density)),
            ),
        )
        val label = measurer.measure("%.2f G".format(peak), arcStyle)
        drawText(
            label,
            topLeft = Offset(cx - label.size.width / 2f, cy - r - label.size.height - 2f * density),
        )
    }

    for ((color, points) in hot) {
        drawPoints(
            points = points,
            pointMode = PointMode.Points,
            color = color.copy(alpha = 0.75f),
            strokeWidth = 4.8f * density,
            cap = StrokeCap.Round,
        )
    }

    // Axis words. The sides say "cornering" rather than left/right, because the
    // side is derived from the steering sign rather than stored.
    val braking = measurer.measure("braking", labelStyle)
    drawText(braking, topLeft = Offset(cx - braking.size.width / 2f, cy - side / 2 - braking.size.height))
    val power = measurer.measure("power", labelStyle)
    drawText(power, topLeft = Offset(cx - power.size.width / 2f, cy + side / 2))
    // The plot is a disc, so a corner of the square is the one place no sample
    // can ever land.
    val cornering = measurer.measure("cornering (G)", labelStyle)
    drawText(cornering, topLeft = Offset(cx - side / 2, cy - side / 2))
}

/**
 * The share of *loaded* samples spent doing two things at once, per highlighted
 * lap, with the session pooled underneath once there is more than one lap to
 * pool — `gripReadoutHtml` in the JS.
 */
@Composable
private fun Readout(
    sg: Grip.SessionGrip,
    rows: List<Pair<Color, Grip.LapShares>>,
    lapNumber: (Int) -> Int,
) {
    if (rows.isEmpty()) return
    val colors = TrackTheme.colors

    @Composable
    fun line(label: @Composable () -> Unit, trail: String, power: String, color: Color) {
        Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            Box(Modifier.weight(1f)) { label() }
            Text(
                trail,
                style = TrackTheme.typography.sm,
                color = color,
                textAlign = TextAlign.End,
                modifier = Modifier.weight(0.8f),
            )
            Text(
                power,
                style = TrackTheme.typography.sm,
                color = color,
                textAlign = TextAlign.End,
                modifier = Modifier.weight(0.8f),
            )
        }
    }

    Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
        line(
            label = {},
            trail = "Braking +\ncornering",
            power = "Cornering +\npower",
            color = colors.textFaint,
        )
        rows.forEach { (color, lap) ->
            line(
                label = {
                    Row(verticalAlignment = Alignment.CenterVertically) {
                        Box(Modifier.size(8.dp).background(color, CircleShape))
                        Text(
                            "  Lap ${lapNumber(lap.chIdx)}",
                            style = TrackTheme.typography.xs,
                            color = colors.textMuted,
                        )
                    }
                },
                trail = "${lap.trailPct.roundToInt()}%",
                power = "${lap.powerPct.roundToInt()}%",
                color = colors.textStrong,
            )
        }
        if (sg.laps.size >= 2) {
            line(
                label = {
                    Text(
                        "Session",
                        style = TrackTheme.typography.xs.copy(fontStyle = FontStyle.Italic),
                        color = colors.textMuted,
                    )
                },
                trail = "${sg.all.trailPct.roundToInt()}%",
                power = "${sg.all.powerPct.roundToInt()}%",
                color = colors.textMuted,
            )
        }
        Text(
            "Share of the samples where the tyre was working (${"%.2f".format(Grip.MIN_LOAD_G)} G combined " +
                "or more) spent doing two things at once. Brake straight, turn, then accelerate and both " +
                "stay low — the gap to the arc is time.",
            style = TrackTheme.typography.xxs,
            color = colors.textFaint,
        )
    }
}
